### PR TITLE
feat(volume/resize): don't allow volume resize in presence of snapshots

### PR DESCRIPTION
**Pull request by dsharma-dc**
If a volume is expanded after taking a snapshot, then the IOs to the new io-units(block ranges) fails because the lba range falls out of bound for replica's parent device(snapshot). Till the issue is fixed, this check ensures we don't try to expand a volume if there is snapshot present. SPDK issue #3108.

### CODE REVIEW

File: control-plane/agents/src/bin/core/volume/operations.rs [hunk 1]
```diff
@@ -191,6 +191,13 @@ impl ResourceResize for OperationGuardArc<VolumeSpec> {
         let spec = self.as_ref().clone();
         let state = registry.volume_state(&request.uuid).await?;
 
+        // Pre-check - Don't allow resize if the volume has snapshots.
+        // NOTE: This is temporary till the fix for handling this in spdk is done.
+        if !registry.specs().snapshots_by_vol(self.uuid()).is_empty() {
+            return Err(SvcError::Internal {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations.rs:197]: @Abhinandan-Purkait will Internal "avoid" the storm of retries or we need a different error code?

**Review by tiagolobocastro — APPROVED** [review]

File: control-plane/agents/src/bin/core/volume/operations.rs (hunk 1 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/operations.rs:197]:
```
codes.Canceled,          // gRPC: Client Application cancelled the request
codes.DeadlineExceeded,  // gRPC: Timeout
codes.Unavailable,       // gRPC: Server shutting down, TCP connection broken - previous volume operation may be still in progress.
codes.ResourceExhausted, // gRPC: Server temporarily out of resources - previous volume operation may be still in progress.
codes.Aborted:           // CSI: Operation pending for volume
```

Only these error codes will do exponential backoff, all others including internal will be burst retries.

Comment (dsharma-dc) [reply control-plane/agents/src/bin/core/volume/operations.rs:197]: `Aborted` may be used here then I guess. Not a perfect fit, but will be ok.

Comment (dsharma-dc) [reply control-plane/agents/src/bin/core/volume/operations.rs:197]: Will map `Aborted` in the CSI controller, let's keep Internal here.

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by dsharma-dc** [comment]
bors merge

**Pull request merged**